internal/ansible/dispatcher/dispatcher_events.go

```diff
@@ -0,0 +1,479 @@
+		Event:       "playbook_on_start",
+		Created:     &event.Play.Duration.Start,
+		RunnerIdent: &d.deviceID,
+		Counter:     msgCounter,
+		EventData: &message.AnsibleRunnerJobEventYamlEventData{
+			PlaybookUuid: &uuid,
+			Playbook:     &playbookFilename,
+			Host:         getHostsString(event.Tasks),
+			Uuid:         &event.Play.Id,
+		},
+		Uuid: uuid,
+	}
+	return &message
+}
+
+func getHostsString(event []results.AnsiblePlaybookJSONResultsPlayTask) *string {
+	hostsString := ""
+	for _, task := range event {
+		hosts := make([]string, 0, len(task.Hosts))
+
+		for k := range task.Hosts {
+			hosts = append(hosts, k)
+		}
+		hostsString = strings.Join(hosts, ",")
+	}
+	return &hostsString
+}
+
+func countOccurence(tasks []results.AnsiblePlaybookJSONResultsPlayTask) map[string]int {
+	dict := make(map[string]int)
+	for _, task := range tasks {
+		dict[task.Task.Id]++
+	}
+	return dict
+}
+
+// required event for cloud connector
+func (d *AnsibleDispatcher) ExecutorOnStart(correlationID string, stdout string) message.AnsibleRunnerJobEventYaml {
+	return message.AnsibleRunnerJobEventYaml{
+		Event:     "executor_on_start",
+		Uuid:      uuid.New().String(),
+		Counter:   -1,
+		Stdout:    stdout,
+		StartLine: 0,
+		EndLine:   0,
+		EventData: &message.AnsibleRunnerJobEventYamlEventData{
+			CrcDispatcherCorrelationId: &correlationID,
+		},
+	}
+}
+
+// required event for cloud connector
+func (d *AnsibleDispatcher) ExecutorOnFailed(correlationID string, stdout string, errorCode string, errorDetails string) message.AnsibleRunnerJobEventYaml {
+	return message.AnsibleRunnerJobEventYaml{
+		Event:     "executor_on_failed",
+		Uuid:      uuid.New().String(),
+		Counter:   -1,
+		Stdout:    stdout,
+		StartLine: 0,
+		EndLine:   0,
+		EventData: &message.AnsibleRunnerJobEventYamlEventData{
+			CrcDispatcherCorrelationId: &correlationID,
+			CrcDispatcherErrorCode:     &errorCode,
+			CrcDispatcherErrorDetails:  &errorDetails,
+		},
+	}
+}
+
+// Create the message with event data to send back to Dispatcher
+/*
+TODO: check if ComposeDispatcherMessage func is correct
+
+return Request('POST', return_url, files={
+        "file": ("runner-events", _newlineDelimited(events), "application/vnd.redhat.playbook.v1+jsonl"),
+        "metadata": "{}"
+    }).prepare()
+
+*/
+func ComposeDispatcherMessage(events []message.AnsibleRunnerJobEventYaml, returnURL string, responseTo string) (*pb.Data, error) {
+
+	content, err := CreateMessage(events)
+	if err != nil {
+		return nil, err
+	}
+
+	httpReq, err := CreateHTTPRequest(returnURL, content)
+	if err != nil {
+		return nil, err
+	}
+
+	bodyContent, err := ioutil.ReadAll(httpReq.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.Data{
+		MessageId:  uuid.New().String(),
+		Directive:  returnURL,
+		ResponseTo: responseTo,
+		Metadata:   map[string]string{"Content-Type": "application/vnd.redhat.playbook.v1+jsonl"}, // TODO: is it correct? In python metadata=req.headers
```

Comment:
Is it correct?

Reply:
No , it wasn't. I updated the code but I'm still investigating on how to retrieve the identity header field.